### ISSUE TITLE
TensorFlowOpLayer messes up the TensorBoard graphs / grouping with TensorBoard

### ISSUE THREAD

**Issue by rani-pinchuk** [tensorflow/tensorflow#39398]
I am not sure if this is a bug, feature request or something I do wrong. I apologize in advance. I do believe this is a non-intended behaviour, or at the very least, a missing documentation. 

As you can see in https://stackoverflow.com/questions/61594352/tensorflowoplayer-messes-up-the-tensorboard-graphs, I have asked this question and get no response. 

I copy the question from Stackoverflow to here:

This question is about TensorFlow (and TensorBoard) version 2.2 (running in tensorflow/tensorflow:2.2.0-gpu-jupyter on Ubuntu 18.04.3 LTS), but I have experienced the same issue with 2.2rc3 and 2.1. It is a continuation of the question '[Messed up TensorBoard graphs due to Python operations](https://stackoverflow.com/questions/61581114/messed-up-tensorboard-graphs-due-to-python-operations)'.

Consider the following code:

```python3
from datetime import datetime
import tensorflow as tf
from tensorflow import keras

inputs = keras.layers.Input(shape=(784, ))    

outputs = tf.zeros([32, 10], tf.float32)

for i in range(0, 3):
    x = keras.layers.Dense(32, activation='relu', name='Model/Block' + str(i) + '/relu')(inputs) 
    x = keras.layers.Dropout(0.2, name='Model/Block' + str(i) + '/dropout')(x)
    x = keras.layers.Dense(10, activation='softmax', name='Model/Block' + str(i) + '/softmax')(x)
    outputs = keras.layers.Lambda(lambda x: x[0] + x[1], name='Model/add/add' + str(i))([outputs, x])

model = tf.keras.Model(inputs=inputs, outputs=outputs)
model.summary(line_length=100, positions=[.45, .58, .67, 1.])

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784).astype('float32') / 255
x_test = x_test.reshape(10000, 784).astype('float32') / 255

model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=keras.optimizers.RMSprop(),
              metrics=['accuracy'])

logdir = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = keras.callbacks.TensorBoard(log_dir=logdir)
model.fit(x_train, y_train,
          batch_size=32,
          epochs=5,
          validation_split=0.2,
          callbacks=[tensorboard_callback])
```

When run, it prints the following summary:
``` 
____________________________________________________________________________________
Layer (type)                          Output Shap Param  Connected to               
====================================================================================
input_1 (InputLayer)                  [(None, 784 0                                 
____________________________________________________________________________________
Model/Block0/relu (Dense)             (None, 32)  25120  input_1[0][0]              
____________________________________________________________________________________
Model/Block0/dropout (Dropout)        (None, 32)  0      Model/Block0/relu[0][0]    
____________________________________________________________________________________
Model/Block1/relu (Dense)             (None, 32)  25120  input_1[0][0]              
____________________________________________________________________________________
Model/Block0/softmax (Dense)          (None, 10)  330    Model/Block0/dropout[0][0] 
____________________________________________________________________________________
Model/Block1/dropout (Dropout)        (None, 32)  0      Model/Block1/relu[0][0]    
____________________________________________________________________________________
Model/Block2/relu (Dense)             (None, 32)  25120  input_1[0][0]              
____________________________________________________________________________________
tf_op_layer_AddV2 (TensorFlowOpLayer) [(32, 10)]  0      Model/Block0/softmax[0][0] 
____________________________________________________________________________________
Model/Block1/softmax (Dense)          (None, 10)  330    Model/Block1/dropout[0][0] 
____________________________________________________________________________________
Model/Block2/dropout (Dropout)        (None, 32)  0      Model/Block2/relu[0][0]    
____________________________________________________________________________________
Model/add/add1 (Lambda)               (32, 10)    0      tf_op_layer_AddV2[0][0]    
                                                         Model/Block1/softmax[0][0] 
____________________________________________________________________________________
Model/Block2/softmax (Dense)          (None, 10)  330    Model/Block2/dropout[0][0] 
____________________________________________________________________________________
Model/add/add2 (Lambda)               (32, 10)    0      Model/add/add1[0][0]       
                                                         Model/Block2/softmax[0][0] 
====================================================================================
```

Note the strange entry `tf_op_layer_AddV2 (TensorFlowOpLayer)`. This kind of entry makes the TensorBoard graphs very messy. It turns out that when avoiding the use of the `tf.zeros()`, this strange `tf_op_layer_AddV2` element is not added. So the following code will not generate any `tf_op_layer` element: 

```python3
from datetime import datetime
import tensorflow as tf
from tensorflow import keras

inputs = keras.layers.Input(shape=(784, ))

x = keras.layers.Dense(32, activation='relu', name='Model/Block0/relu')(inputs) 
x = keras.layers.Dropout(0.2, name='Model/Block0/dropout')(x)
outputs = keras.layers.Dense(10, activation='softmax', name='Model/Block0/softmax')(x)

for i in range(1, 3):
    x = keras.layers.Dense(32, activation='relu', name='Model/Block' + str(i) + '/relu')(inputs) 
    x = keras.layers.Dropout(0.2, name='Model/Block' + str(i) + '/dropout')(x)
    x = keras.layers.Dense(10, activation='softmax', name='Model/Block' + str(i) + '/softmax')(x)
    outputs = keras.layers.Lambda(lambda x: x[0] + x[1], name='Model/add/add' + str(i))([outputs, x])

model = tf.keras.Model(inputs=inputs, outputs=outputs)
model.summary(line_length=84, positions=[.46, .60, .69, 1.])

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784).astype('float32') / 255
x_test = x_test.reshape(10000, 784).astype('float32') / 255

model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=keras.optimizers.RMSprop(),
              metrics=['accuracy'])

logdir = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = keras.callbacks.TensorBoard(log_dir=logdir)
model.fit(x_train, y_train,
          batch_size=32,
          epochs=5,
          validation_split=0.2,
          callbacks=[tensorboard_callback])

```
There are many more complex examples where the `tf_op_layer_` elements are created. It will be appreciated if it is exaplained why they are created and how to avoid them.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/b2104ca83ce67e0789c4fa912a2bc624/39398.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/f8d53f50bc7dfceae9e31d93f88204e3/39398-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by jvishnuvardhan** [comment]
@rani-pinchuk Just want to know you use-case with more details. Why do we have two things (`outputs` layer and `outputs` tensor) in the model? What is your intentions in defining two with the same name? I think that resulted the complicated model. Can you change `outputs` tensor to some other name and define you model. Thanks!

**Comment by rani-pinchuk** [comment]
My use case is to get a useful Tensorboard graph for a large model (which I am afraid I cannot share). In that model, I use different operations like '+' or t[begin:end] and these operations mess up the model due to the prefixes created by TensorFlowOpLayer. 

The example above is just an example to reproduce this behaviour. I have more examples where this is happening, for example with the t[begin:end] operation. Also shape operations create this problem. 

In the low level TensorFlow 1.0 it was possible to group operations and therefore generate very nice graphs. It seems that in TensorFlow 2.0 with Keras, the way to do that is by using name hierarchy (e.g. 'Model/add/add') (I could not find any documentation about how to group operations in TF2 and found about the name hierarchy trick from examples online, so I am not sure that I don't mis anything here). But the TensorFlowOpLayer messes this up. 

Regarding the `outputs` Tensor in the example - there is one `outputs` identifier. There is no `outputs` layer. The `outputs` identifier is first used to hold a Tensor of zeros. In the loop, the `outputs` identifier gets the sum of the last Tensor referred by that identifier and the result from the Dense and Dropout layers. It could be written more clearly like this:
```python
outputs = tf.zeros([32, 10], tf.float32)

for i in range(0, 3):
    x = keras.layers.Dense(32, activation='relu', name='Model/Block' + str(i) + '/relu')(inputs) 
    x = keras.layers.Dropout(0.2, name='Model/Block' + str(i) + '/dropout')(x)
    x = keras.layers.Dense(10, activation='softmax', name='Model/Block' + str(i) + '/softmax')(x)
    outputs = outputs + x
```
However, in the above (which is the origin of whole this problem) the `+` operation is unnamed, and therefore will mess the TensorBoard graph even more (see https://stackoverflow.com/questions/61581114/messed-up-tensorboard-graphs-due-to-python-operations/61594411). 

BTW, I have tried to change the `outputs` Tensor name (see https://stackoverflow.com/questions/61594352/tensorflowoplayer-messes-up-the-tensorboard-graphs). This had no effect.

**Comment by nfelt** [comment]
This looks more like an issue with the naming of the TensorFlowOpLayer as generated within Keras than an issue with TensorBoard per se; if the layer name discards the hierarchy information we can't really recreate it.

Assigning to @omalleyt12 to assess since it looks like you wrote TensorFlowOpLayer?

cc @davidsoergel as current TensorBoard onduty

**Comment by saikumarchalla** [comment]
@rani-pinchuk Seems the issue is fixed in latest TF version(2.5 and Nightly) and `tf_op_layer_AddV2 (TensorFlowOpLayer)` is not getting added in the model summary. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/23cfb01b7ad8a9f7d0aa05d35ce43b86/untitled79.ipynb).

Closing this issue as of now. Please feel free to re-open the issue if you any concern.
